### ISSUE TITLE
iOS and tvOS 15 support

### ISSUE THREAD

**Issue by defagos** [SRGSSR/playsrg-apple#180]
Here is the list of the issues found with the current betas:

- [x] Accessibility via VoiceOver does not work at all on tvOS for SwiftUI views. (ℹ️: We [reported](https://openradar.appspot.com/FB9524280) this issue to Apple. Not fixed in the RC 😭)
- [x] On tvOS the tab bar custom font is initially incorrect (fixed after focusing a tab). This was working fine on tvOS 14. (✅: Works fine when compiled with the tvOS 15 SDK and run on tvOS 15).
- [x] Compilation of Letterbox does not work, see https://github.com/SRGSSR/srgletterbox-apple/issues/249 (✅: Works with Xcode 13b5).
- [x] SwiftMessages does not compile (https://github.com/SwiftKickMobile/SwiftMessages/issues/471) (✅: Use 9.0.4 pre-release branch).
- [x] For binaries compiled with the tvOS 14 SDK (thus with the old tvOS player experience), opening and closing the info panel, then exiting the player, keeps the player playing sound in the background for ~30 s. This is not related to an SRG Media Player or Letterbox timer, but probably to an AVKit internal timer. The issue namely does not occur when running on a tvOS 14 device, and if we compile the same app with the tvOS 15 SDK to run it on a tvOS 15 device. But we should at least check that this stays the same with the tvOS 15 RC.
- [x] The navigation bar is broken on iOS (radio tab, song panel) (✅: Fixed with modern navigation bar API, see #151)
- [x] The legacy Show A-Z view crashes (✅: This view has been replaced).
- [x] The tab bar blur sometimes breaks (e.g. when pushing the TV guide view). iOS 15 [hides the blur](https://developer.apple.com/forums/thread/682432) when there is no content behind (can be tested with a few favorites, not filling the screen; when at rest there is no blur, but if we move a few cells in the vicinity of the tab bar the blur appears; can also be tested with the App Store app). We need to update our code to take into account this new behavior. (✅: Fixed with proper use of the appearance API, as this is an expected change; the new iOS 15 look & feel at scroll edge has not been adopted for the time being, though).
- [x] When returning back after tapping on a focused item, the item stays focused. Our companion text animates, which now produces a weird result (ℹ️: We cannot do much, focus state is communicated in the same way by the system as before, its is only images in buttons that behave differently)
- [x] Assertion failure on the program guide when attempting to scroll to a location outside the visible range, e.g. 5 PM. As a result the table jumps, something not visible with a nightly built with iOS 14 and run on 15. Probably related to the new cell prefetch mechanism. The relevant error is:

    ```
    *** Assertion failure in -[UICollectionView _validateScrollingTargetIndexPath:], UICollectionView.m:6485
    2021-08-20 17:32:02.174290+0200 Play RTS[1093:151623] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Attempted to scroll the collection view to an out-of-bounds item (25) when there are only 0 items in section 0. Collection view: <UICollectionView: 0x1072d0000; frame = (10 0; 355 423); clipsToBounds = YES; gestureRecognizers = <NSArray: 0x283f6ce40>; layer = <CALayer: 0x283461f60>; contentOffset: {0, 0}; contentSize: {355, 3}; adjustedContentInset: {0, 0, 55, 0}; layout: <UICollectionViewCompositionalLayout: 0x104eccd80>; dataSource: <_TtGC5UIKit34UICollectionViewDiffableDataSourceOC7PlaySRG26ProgramGuideDailyViewModel7SectionCSo10SRGProgram_: 0x2831663d0>>.'
    ```

Also see https://github.com/SRGSSR/srgmediaplayer-apple/issues/104 and https://github.com/SRGSSR/srgletterbox-apple/issues/249 (✅: Fixed)

We should also use new APIs available since iOS and tvOS 15:

- [x] [Blurred views](https://www.hackingwithswift.com/quick-start/swiftui/how-to-add-visual-effect-blurs) which should eliminate header background animation glitches (✅: Implemented)
- [x] Livestream metadata in tvOS 15 control center (ℹ️: Does not seem to work in the RC; we will keep our live label for the time being)

**Comment by defagos** [comment]
We should also check whether cell prefetching does not [collide](https://twitter.com/steipete/status/1420722411629056003?s=21) with our (small) usage of visible cells.

**Comment by defagos** [comment]
Compilation issues with SPM are a documented change and we probably have to do something, as said in the [Xcode 13 beta release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-13-beta-release-notes):

> Linking Swift packages from application extension targets or watchOS applications no longer emits unresolvable warnings about linking to libraries not safe for use in application extensions. This means that code referencing APIs annotated as unavailable for use in app extensions must now themselves be annotated as unavailable for use in application extensions, in order to allow that code to be used in both apps and app extensions. (66928265)

Airship delivered [version 14.6.0 to fix this issue](https://github.com/urbanairship/ios-library/issues/286).

**Comment by defagos** [comment]
Work started on `feature/ios15-tvos15-xcode13`.

**Comment by defagos** [comment]
The navigation bar issue is likely due to the use of tricks. I remember Marco Arment talking about such issues with Overcast on iOS 15. Fortunately migration to the iOS 13 API fixes the issue, and we already did it on a feature branch (#151). I'll cherry-pick this implementation on the branch associated with the present issue.

**Comment by defagos** [comment]
We did everything we could, reported remaining issues to Apple, and identified other behaviors as part of OS changes which we must accept.

**Comment by hansemannn** [comment]
Hi @defagos, sorry to slide into this issue! But may you still know how you fixed the "Attempted to scroll the collection view to an out-of-bounds item" error? It seems to happen randomly in one of my test projects and the cell structure looks fine. Thanks in advance!

**Comment by defagos** [comment]
Hi @hansemannn. Sure, I do remember. We use a small Objective-C [method wrapper](https://github.com/SRGSSR/playsrg-apple/blob/ios/3.5.0-373/Application/Sources/UI/Helpers/UICollectionView+PlaySRG.m) to trap the exception that may be thrown by the scrolling method.
